### DISCUSSION TITLE
Suggestion: Exclude patterns GUI improvements

### DISCUSSION THREAD

**Discussion by geckolinux** [borgbase/vorta#838] [Ideas]
First of all, many thanks for Vorta, this is an extremely important tool to motivate less technical people to take advantage of the all the goodness of Borg.

The only part of the GUI that I found slightly confusing was the exclude patterns, especially after reading the link to the manual, which for a non-technical user doesn't explain clearly what exclude patterns are for or how to use them and how they differ from the other apparently similar methods for excluding certain files and/or directories from backups.

I think it would be nice for Vorta to have a few GUI exclusion improvements, while still allowing to manually type in patterns for experts who prefer that. Maybe something like:

1. A field with directories and files to be excluded, basically the inverse of the **Source Folders and Files to Back Up** pane. This will probably cover 95% of beginners' needs.
2. A checkbox to exclude folders with the `.nobackup` file.
3. A GUI way of building slightly more complex exclusion rules, similar to buildregex.com , but obviously not as comprehensive, just to cover the most common rules of files and/or directories that start with `BLAH`, that contain`BLAH` anywhere in the name, or end with `BLAH`, with a little additional help on the use of wildcards.

Thanks a lot!

**Comment by m3nu** [comment]
I like the checkbox idea, item 2.

We probably don't have the resources to build item 3, but I'd like to have some examples for different OS.

Item 1 would be a list of absolute paths. Also possible, I think.

**Comment by m3nu** [comment]
Just noticed that Nextcloud (desktop client) comes with a nice list of default exclusions. Maybe something for Vorta to assimilate.

<img width="500" src="https://user-images.githubusercontent.com/3916435/85429630-9745e300-b5b1-11ea-9792-8e376b7f7ed6.jpg" />


I agree that one simple kind of patterns would be enough.

**Comment by m3nu** [comment]
Found some more details on macOS exclusions. They work via extended attributes. I think excluding those should be implemented in Borg directly. It already processes xattrs.

https://apple.stackexchange.com/questions/25779/on-os-x-what-files-are-excluded-by-rule-from-a-time-machine-backup

**Comment by mauromol** [comment]
I also found the exclusions part of the GUI quite confusing.

First of all, it's not really clear what type of exclusion syntax should I use there: the link to documentation sends to borg documentation where ALL exclusion syntaxes are explained. So, it's not clear whether I should assume Vorta passes my exclusion rules as -e/--exclude/--exclude-from or as --pattern/--patterns-from, so whether default style is fm: or rather sh: for instance.

Partly because of this and partly because of borg itself, if I add "/home/user/mauro" as source and I specify "cache" as an exclusion rule, I would expect "/home/user/mauro/cache" to be excluded (that is: "cache" being relative to the source), but it's not the case, I must specify "/home/user/mauro/cache". As I said, this is not completely Vorta's fault, however I think that a good way to workaround this would be to provide a preview button and dialog where I can see which files will be excluded based on my current rules. I don't know if this can be done by issuing a "borg create" command with --dry-run and proper logging settings: it would be extremely useful for me!

Just an additional note: this would be even more useful due to the fact that if you proceed by trial-and-error instead, the cancel button often makes Vorta freeze due to #332 😜

**Comment by m3nu** [comment]
Yeah, this tab will be redone eventually.

In the meanwhile feel free to suggest new concepts or [screenshots](https://github.com/borgbase/vorta/issues/399#issuecomment-583844419) that could be useful here. I.e. to make Borg's different exclusion options accessible from a GUI.

**Comment by upward4** [comment]
Is this supposed to leverage `fnmatch` under the hood? Because it doesn't seem to be working. For example, I am trying to use `[!/]` inside an exclude pattern, but it seems to be ignored.

Example: ignore `/home/user/Downloads` but *not* `/home/user/foo/Downloads`:

```
/home/[!/]/Downloads
```

EDIT: I think I misunderstood how `[seq]` works. Indeed, it is supposed to match a single character, not a string. Does anyone know how to deal with the above example using `fnmatch` syntax?

**Reply by m3nu** [reply]
YES! Vorta passes your excludes as `--exclude-from`, so everything mentioned in the [docs](https://borgbackup.readthedocs.io/en/stable/usage/help.html#borg-help-patterns) applies, including the use of [fnmatch](https://docs.python.org/3/library/fnmatch.html).

**Comment by jcpsantiago** [comment]
+1 here to make it friendlier 😬 I'm trying Vorta for the first time today (and Borg for that matter) and I'm confused with the exclusion patterns too. if I add `*/Library` I expect `/Users/santiago/Library` to be excluded and the size of the backup to go down after I press `Recalculate sizes`, but it stays the same. Is my assumption wrong?

**Reply by m3nu** [reply]
*Recalculate Sizes* is a very simple function inside Vorta and doesn't know about excludes.

To exclude the `Library` folder, you could use `~/Library`, since we expand `~`.

`*/Library` will also exclude subfolders with that name, which you may not want. Something like `*/.git` would be a more common use case.

**Reply by jcpsantiago** [reply]
ah that makes sense, thought "size" represented what would, in fact, be stored i.e. a dry-run estimate. Thanks for clarifying that.

**Comment by thibaultmol** [comment]
> Recalculate Sizes is a very simple function inside Vorta and doesn't know about excludes.

As a user, I would have expected it to look at excludes (that way you can kinda check if your excludes are working correctly.

FR: having a way to check if a file would be included or not (based on these patterns) would be nice. Cause for most people this will be a black box in a sense

**Comment by real-yfprojects** [comment]
> 3. A GUI way of building slightly more complex exclusion rules, similar to buildregex.com , but obviously not as comprehensive, just to cover the most common rules of files and/or directories that start with `BLAH`, that contain`BLAH` anywhere in the name, or end with `BLAH`, with a little additional help on the use of wildcards.

I can't find a GUI for constructing regexes on the website mentioned ([buildregex.com](https://buildregex.com)).
